File: services/2-redis-cache/go/internal/model/similarity.go
// Пакет model - модель сходства фильмов в оперативной памяти.
//
// Загружается один раз при старте сервиса через тот же пул pgxpool,
// что и обычные запросы.
package model

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/jackc/pgx/v5/pgxpool"
)

type Neighbor struct {
	MovieID    int
	Similarity float64
}

type SimilarityModel struct {
	mu          sync.RWMutex
	data        map[int][]Neighbor // movie_id -> отсортированные соседи
	loaded      bool
	loadTimeSec float64
}

func New() *SimilarityModel {
	return &SimilarityModel{
		data: make(map[int][]Neighbor),
	}
}

func (m *SimilarityModel) Load(ctx context.Context, pool *pgxpool.Pool) error {
	start := time.Now()

	rows, err := pool.Query(ctx,
		"SELECT movie_id, similar_movie_id, similarity FROM item_similarity ORDER BY movie_id, similarity DESC")
	if err != nil {
		return fmt.Errorf("запрос: %w", err)
	}
	defer rows.Close()

	data := make(map[int][]Neighbor)
	count := 0
	for rows.Next() {
		var movieID, similarID int
		var sim float64
		if err := rows.Scan(&movieID, &similarID, &sim); err != nil {
			return fmt.Errorf("чтение строки: %w", err)
		}
		data[movieID] = append(data[movieID], Neighbor{MovieID: similarID, Similarity: sim})
		count++
	}
	if err := rows.Err(); err != nil {
		return fmt.Errorf("перебор строк: %w", err)
	}

	m.mu.Lock()
	m.data = data
	m.loaded = true
	m.loadTimeSec = time.Since(start).Seconds()
	m.mu.Unlock()

	log.Printf("Модель загружена: %d фильмов, %d пар, %.1fс", len(data), count, m.loadTimeSec)
	return nil
}

func (m *SimilarityModel) GetNeighbors(movieID int) []Neighbor {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.data[movieID]
}

func (m *SimilarityModel) HasMovie(movieID int) bool {
	m.mu.RLock()
	defer m.mu.RUnlock()
	_, ok := m.data[movieID]
	return ok
}

func (m *SimilarityModel) IsLoaded() bool {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.loaded
}

func (m *SimilarityModel) LoadTime() float64 {
	m.mu.RLock()
	defer m.mu.RUnlock()
	return m.loadTimeSec
}

File: services/2-redis-cache/python/app/similarity.py
"""
Загрузка модели сходства фильмов из БД в оперативную память.

Модель: dict[movie_id -> list[(similar_movie_id, similarity)]]
Загружается один раз при старте сервиса (через тот же пул asyncpg,
что и обычные запросы).
"""

from __future__ import annotations

import time
from collections import defaultdict

from app import database


# Глобальная модель: {movie_id: [(similar_movie_id, similarity), ...]}
_model: dict[int, list[tuple[int, float]]] = {}
_model_loaded = False
_load_time_seconds = 0.0


async def load_model() -> None:
    """Загружает item_similarity из PostgreSQL в оперативную память."""
    global _model, _model_loaded, _load_time_seconds

    start = time.monotonic()
    pool = database.get_pool()
    rows = await pool.fetch(
        "SELECT movie_id, similar_movie_id, similarity "
        "FROM item_similarity ORDER BY movie_id, similarity DESC"
    )
    model: dict[int, list[tuple[int, float]]] = defaultdict(list)
    count = 0
    for row in rows:
        model[row["movie_id"]].append((row["similar_movie_id"], row["similarity"]))
        count += 1

    _model = dict(model)
    _model_loaded = True
    _load_time_seconds = time.monotonic() - start
    print(
        f"Модель загружена: {len(_model)} фильмов, "
        f"{count} пар, {_load_time_seconds:.1f}с"
    )


def get_neighbors(movie_id: int) -> list[tuple[int, float]]:
    """Возвращает K ближайших соседей фильма."""
    return _model.get(movie_id, [])


def is_loaded() -> bool:
    return _model_loaded


def load_time() -> float:
    return _load_time_seconds
